server/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportResizeAction.java

```diff
@@ -129,12 +138,37 @@ static CreateIndexClusterStateUpdateRequest prepareCreateIndexRequest(final Resi
         targetIndexSettingsBuilder.remove(IndexMetadata.SETTING_HISTORY_UUID);
         final Settings targetIndexSettings = targetIndexSettingsBuilder.build();
         final int numShards;
+        ByteSizeValue maxSinglePrimarySize = resizeRequest.getMaxSinglePrimarySize();
         if (IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.exists(targetIndexSettings)) {
+            if (resizeRequest.getResizeType() == ResizeType.SHRINK && maxSinglePrimarySize != null) {
+                throw new IllegalArgumentException("Cannot set both index.number_of_shards and max_single_primary_size" +
+                    " for the target index");
+            }
             numShards = IndexMetadata.INDEX_NUMBER_OF_SHARDS_SETTING.get(targetIndexSettings);
         } else {
             assert resizeRequest.getResizeType() != ResizeType.SPLIT : "split must specify the number of shards explicitly";
             if (resizeRequest.getResizeType() == ResizeType.SHRINK) {
-                numShards = 1;
+                if (maxSinglePrimarySize != null) {
+                    int sourceIndexShardsNum = sourceMetadata.getNumberOfShards();
+                    long sourceIndexStorageBytes = indexStoreStats.getSizeInBytes();
+                    long maxSinglePrimarySizeBytes = maxSinglePrimarySize.getBytes();
+                    long minShardsNum = sourceIndexStorageBytes / maxSinglePrimarySizeBytes;
+                    if (minShardsNum * maxSinglePrimarySizeBytes < sourceIndexStorageBytes) {
+                        minShardsNum = minShardsNum + 1;
+                    }
+                    if (minShardsNum > sourceIndexShardsNum) {
+                        logger.info("By setting max_single_primary_size to [" + maxSinglePrimarySize.toString() +
+                            "], the target index [" + targetIndexName + "] will contain [" + minShardsNum +
+                            "] shards, which will be greater than [" + sourceIndexShardsNum +
+                            "] shards of the source index [" + sourceMetadata.getIndex().getName() +
+                            "], use [" + sourceIndexShardsNum + "] as the shards number of the target index [" + targetIndexName + "]");
```

Comment:
This can be cleaned up using the `{}` substitution

```suggestion
                        logger.info("By setting max_single_primary_size to [{}], the target index [{}] will contain [{}] shards, which will be greater than [{}] shards in the source index [{}], using [{}] for the shard count of the target index [{}]",
                            maxSinglePrimarySize.toString(), targetIndexName, minShardsNum, sourceIndexShardsNum, sourceMetadata.getIndex().getName(), sourceIndexShardsNum, targetIndexName);
```